controllers/librarypanel_controller.go

```diff
@@ -0,0 +1,432 @@
+
+	return ctrl.Result{RequeueAfter: libraryPanel.Spec.ResyncPeriod.Duration}, nil
+}
+
+func (r *GrafanaLibraryPanelReconciler) reconcileWithInstance(ctx context.Context, instance *v1beta1.Grafana, cr *v1beta1.GrafanaLibraryPanel, modelWithHash *libraryPanelModelWithHash) error {
+	if instance.IsInternal() {
+		err := ReconcilePlugins(ctx, r.Client, r.Scheme, instance, cr.Spec.Plugins, fmt.Sprintf("%v-librarypanel", cr.Name))
+		if err != nil {
+			return err
+		}
+	} else if instance.IsExternal() && cr.Spec.Plugins != nil {
+		return fmt.Errorf("external grafana instances don't support plugins, please remove spec.plugins from your library panel cr")
+	}
+
+	grafanaClient, err := client2.NewGeneratedGrafanaClient(ctx, r.Client, instance)
+	if err != nil {
+		return err
+	}
+
+	folderUID, err := getFolderUID(ctx, r.Client, cr)
+	if err != nil {
+		return err
+	} else if folderUID == "" {
+		// this indicates an implementation error; we should be returning errors
+		// if the folder isn't found, but an empty UID is still technically possible to return.
+		return fmt.Errorf("found folder but could not determine its UID")
+	}
+
+	uid := content.CustomUIDOrUID(cr, fmt.Sprintf("%s", modelWithHash.Model["uid"]))
+	name := fmt.Sprintf("%s", modelWithHash.Model["name"])
+
+	desired := &libraryPanelToReconcile{
+		Name:          name,
+		FolderUID:     folderUID,
+		Kind:          int64(libraryElementTypePanel),
+		UID:           uid,
+		ModelWithHash: modelWithHash,
+		Version:       0, // this will be filled in by computeOperation in case of update
+	}
+
+	op, err := r.computeOperation(grafanaClient, cr, desired)
+	if err != nil {
+		return err
+	}
+
+	switch op {
+	case libraryElementOperationNoop: // do nothing
+	case libraryElementOperationCreate:
+		// nolint:errcheck
+		_, err = grafanaClient.LibraryElements.CreateLibraryElement(&models.CreateLibraryElementCommand{
+			FolderUID: desired.FolderUID,
+			Kind:      desired.Kind,
+			Model:     desired.ModelWithHash.Model,
+			Name:      desired.Name,
+			UID:       desired.UID,
+		})
+		if err != nil {
+			return err
+		}
+	case libraryElementOperationUpdate:
+		// nolint:errcheck
+		_, err = grafanaClient.LibraryElements.UpdateLibraryElement(desired.UID, &models.PatchLibraryElementCommand{
+			FolderUID: desired.FolderUID,
+			Kind:      desired.Kind,
+			Model:     desired.ModelWithHash.Model,
+			Name:      desired.Name,
+			Version:   desired.Version,
+		})
+		if err != nil {
+			return err
+		}
+	}
+
+	instance.Status.LibraryPanels = instance.Status.LibraryPanels.Add(cr.Namespace, cr.Name, uid)
+	return r.Client.Status().Update(ctx, instance)
+}
+
+// computeOperation looks at the current state of Grafana versus the CR and determines
+// whether a create or update operation should take place (or neither.)
+func (r *GrafanaLibraryPanelReconciler) computeOperation(client *genapi.GrafanaHTTPAPI, cr *v1beta1.GrafanaLibraryPanel, desired *libraryPanelToReconcile) (libraryElementOperation, error) {
```

Comment:
I think it would be cleaner to get rid of this function and have the functionality in `reconcileWithInstance` instead